core/templates/cowdata.h

```diff
@@ -269,22 +269,28 @@ void CowData<T>::_unref() {
 
 	SafeNumeric<USize> *refc = _get_refcount();
 	if (refc->decrement() > 0) {
-		return; // still in use
+		// Data are still in use elsewhere.
+		_ptr = nullptr;
+		return;
 	}
-	// clean up
+	// Clean up.
+	// First, invalidate our own reference.
+	// If we are involved in a cyclic reference, we may otherwise accidentally
+	//  expose the buffer even though it's in the middle of being destructed.
+	// This is illegal behavior anyway, but we can mitigate the damage by
+	//  invalidating our own reference before calling the destructors.
+	USize current_size = *_get_size();
+	T *ptr = _ptr;
```

Comment:
Maybe rename this as tmp_ptr or something, `ptr` vs `_ptr` is going to cause someone to make a mistake I think

Reply:
Perfect, thank you.